### PR TITLE
Terrain estimation when height estimation is based on range finder data

### PR DESCRIPTION
**Pull request by ecmnet**
When height estimation is based on range finder data, the terrain estimation must not use a range finder as source. Either terrain estimation needs to be stopped or uses a different source. This is an open PR to find an appropriate and generic solution to this issue. 

As a first step, terrain estimation is paused as soon as a fallback from vision based to range finder based height estimation occurs. Terrain estimation is continued when vision data is available again. However, this workaround is not generic enough as it only covers the specific use case described.

Refer to https://github.com/PX4/PX4-ECL/pull/994#issuecomment-810309127

### CODE REVIEW

File: EKF/ekf.cpp [hunk 1]
```diff
@@ -119,8 +119,9 @@ bool Ekf::update()
 		// control fusion of observation data
 		controlFusionModes();
 
-		// run a separate filter for terrain estimation
-		runTerrainEstimator();
+		// run a separate filter for terrain estimation if not in vision fallback with rangefinder
+		if(!(_control_status.flags.rng_hgt && _params.vdist_sensor_type == VDIST_SENSOR_EV))
```
Comment (dagar) [inline EKF/ekf.cpp:123]: Can you add braces?

Comment (dagar) [inline EKF/ekf.cpp:123]:
```suggestion
		if(!(_control_status.flags.rng_hgt && _params.vdist_sensor_type == VDIST_SENSOR_EV)) {
```

Comment (ecmnet) [reply EKF/ekf.cpp]: Sure

**Comment by ecmnet** [comment]
To avoid drift during fallbacks, terrain estimator should not run for modes where a valid estimation is not possible:

- Rangefinder based height 
- intermediate states where height estimation purely relies on ACC

**Comment by ecmnet** [comment]
@mhkabir, @bresch : Your comments?

**Comment by priseborough** [comment]
I agree that we should not be running the terrain estimator if ground is being used as the height datum for the EKF, ie using range finder. Do you have test evidence that this fixes the issue reported? The change looks sensible.

**Comment by ecmnet** [comment]
The change is tested in the context of vision as primary source to survive reset cycles of the vision subsystem.

**Comment by ecmnet** [comment]
Further tests have shown that this does not work in the case when vision fails before terrain is estimated the first time (initialization?). Any suggestions how to solve this?

**Comment by priseborough** [comment]
> Further tests have shown that this does not work in the case when vision fails before terrain is estimated the first time (initialization?). Any suggestions how to solve this?

@ecmnet Are you able to provide the steps you used to create the input data that resulted in the failure case? I would like to make a unit test for corner cases like this.

**Comment by ecmnet** [comment]
I tested this with a live vehicle using odometry and a rangefinder with propellers unmounted. After simulating a takeoff and waiting for valid terrain estimation, I switched off vision observing z estimation following raw lidar data.

**Comment by dagar** [comment]
ECL has been absorbed into PX4-Autopilot, can you recreate this PR over there?

https://github.com/PX4/PX4-Autopilot/tree/master/src/lib/ecl

**Comment by ecmnet** [comment]
Will do as soon as I get a running version based on v1.12.0. Currently I have the issue, that 'freezing' the ground level takes place too late (about 5secs). I also want to adjust initialisation in the case that the initial primary height source is a range finder .

**Comment by ecmnet** [comment]
> ECL has been absorbed into PX4-Autopilot, can you recreate this PR over there?
> 
> https://github.com/PX4/PX4-Autopilot/tree/master/src/lib/ecl

Link is not working

**Comment by dagar** [comment]
@ecmnet I initially merged it into `src/lib/ecl` (where the submodule lived), then later moved it to within `src/modules/ekf2/EKF`.

https://github.com/PX4/PX4-Autopilot/tree/master/src/modules/ekf2/EKF

**Comment by ecmnet** [comment]
Further tests show, that there is a dependency to `Ekf::controlHeightSensorTimeouts()`: Within the phase the selected height sensor idoes not provide any data and the detection of the timeout, terrain estimation is continued and relies e.g. on RF. As no height source is available, Z position is considered to be (more or less) constant. If the vehicle actually moves in Z direction during this phase, this movement is falsely reflected as movement of the terrain altitude. With the detection of the timeout, the terrain altitude is freezed. The freezed altitude in this case is not correct, depending on the selected timeout time and the speed of the vehicle in Z-direction.

Any suggestions how to solve this?

**Comment by priseborough** [comment]
There may reasons to use range finder for terrain offset and EKF height (assuming terrain is at zero) simultaneously due to differing requirements for filtering out the effect of terrain height variation. We will move this to the PX4-Autopilot issues list and continue the investigation.
